src/function.ts

```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
   context: Context
 ) => Promise<O>;
 
-export interface IFunction<P, O> {
+/**
+ * Returns the payload type on the {@link IFunction}.
+ */
+export type FunctionPayload<F extends IFunction<any, any>> = [F] extends [IFunction<infer P, any>] ? P : never;
+/**
+ * Returns the output type on the {@link IFunction}.
+ */
+ export type FunctionOutput<F extends IFunction<any, any>> = [F] extends [IFunction<any, infer O>] ? O : never;
+
+type FunctionAsyncOnFailureDestination<P> =
+  | aws_lambda.FunctionProps["onFailure"]
+  | IEventBus<AsyncResponseFailureEvent<P>>
+  | IFunction<AsyncResponseFailure<P>, any>;
+
+type FunctionAsyncOnSuccessDestination<P, O> =
+  | aws_lambda.FunctionProps["onSuccess"]
+  | IEventBus<AsyncResponseSuccessEvent<P, O>>
+  | IFunction<AsyncResponseSuccess<P, O>, any>;
+
+/**
+ * Wrapper around {@link aws_lambda.EventInvokeConfigOptions} which allows users to provide Functionless
+ * {@link EventBus} and {@link Function} for the onSuccess and onFailure async event destinations.
+ */
+export interface EventInvokeConfigOptions<P, O>
+  extends Omit<aws_lambda.EventInvokeConfigOptions, "onSuccess" | "onFailure"> {
+  onSuccess?: FunctionAsyncOnSuccessDestination<P, O>;
+  onFailure?: FunctionAsyncOnFailureDestination<P>;
+}
+
+export interface IFunction<in P, O> {
   readonly functionlessKind: typeof Function.FunctionlessType;
   readonly kind: typeof Function.FunctionlessType;
   readonly resource: aws_lambda.IFunction;
 
   (...args: Parameters<ConditionalFunction<P, O>>): ReturnType<
     ConditionalFunction<P, O>
   >;
+
+  /**
+   * Event Source for the {@link Function} onSuccess async invocation destination.
+   *
+   * For Lambda, the onSuccess destination must first be configured via either
+   * the {@link Function} constructor or by using {@link IFunction.configureAsyncInvoke}
+   * and that destination must match the bus provided here.
+   *
+   * @see https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html#invocation-async-destinations
+   *
+   * The rule returned will contain the logic:
+   *
+   * ```ts
+   * when(id, event => event.source === "lambda"
+   *    && event["detail-type"] === "Lambda Function Invocation Result - Success"
+   *      && event.resources.includes(this.resource.functionArn))
+   * ```
+   */
+  onSuccess(
```

Comment:
nit: looks like eslint missed this missing `public`

Reply:
Thats the interface :-)

Reply:
Ahhhh my bad. Thumbs down feels bad lol